lib/ed25519.js

```diff
@@ -0,0 +1,148 @@
+ * @throws {TypeError} Throws if the supplied buffer is not of the right size
+ *  or not a Uint8Array or Buffer.
+ *
+ * @returns {Buffer} DER private key prefix + key bytes.
+*/
+function privateKeyDerEncode({privateKeyBytes, seedBytes}) {
+  if(!(privateKeyBytes || seedBytes)) {
+    throw new TypeError('`privateKeyBytes` or `seedBytes` is required.');
+  }
+  if(!privateKeyBytes && !(seedBytes instanceof Uint8Array &&
+    seedBytes.length === 32)) {
+    throw new TypeError('`seedBytes` must be a 32 byte Buffer.');
+  }
+  if(!seedBytes && !(privateKeyBytes instanceof Uint8Array &&
+    privateKeyBytes.length === 64)) {
+    throw new TypeError('`privateKeyBytes` must be a 64 byte Buffer.');
+  }
+  let p;
+  if(seedBytes) {
+    p = seedBytes;
+  } else {
+    // extract the first 32 bytes of the 64 byte private key representation
+    p = privateKeyBytes.slice(0, 32);
+  }
+  return Buffer.concat([DER_PRIVATE_KEY_PREFIX, p]);
+}
+
+/**
+ * Takes a Buffer of public key bytes and encodes it in DER-encoded
+ * SubjectPublicKeyInfo (SPKI) format.
+ * Allows Uint8Arrays to be interoperable with node's crypto functions.
+ *
+ * @param {object} options - Options to use.
+ * @param {Uint8Array} options.publicKeyBytes - The keyBytes.
+ *
+ * @throws {TypeError} Throws if the bytes are not Uint8Array or of length 32.
+ *
+ * @returns {Buffer} DER Public key Prefix + key bytes.
+*/
+function publicKeyDerEncode({publicKeyBytes}) {
+  if(!(publicKeyBytes instanceof Uint8Array && publicKeyBytes.length === 32)) {
+    throw new TypeError('`publicKeyBytes` must be a 32 byte Buffer.');
+  }
+  return Buffer.concat([DER_PUBLIC_KEY_PREFIX, publicKeyBytes]);
+}
+
+// seedBytes can be a TypedArray or a function that returns a TypedArray
+const getSeedBytes = async seed => {
+  if(typeof seed === 'function') {
+    return seed();
+  }
+  return seed;
+};
+
+const api = {
+  async generateKeyPairFromSeed(seed) {
+    const seedBytes = await getSeedBytes(seed);
+    const publicBytes = new Uint8Array(32);
+    const privateBytes = new Uint8Array(64);
+    privateBytes.set(seedBytes);
+    privateBytes.set(publicBytes, 32);
+    const privateKey = await createPrivateKey({
+      // node is more than happy to create a new private key using a DER
+      key: privateKeyDerEncode({privateKeyBytes: privateBytes}),
+      format: 'der',
+      type: 'pkcs8'
+    });
+    // this expects either a PEM encoded key or a node privateKeyObject
+    const publicKey = await createPublicKey(privateKey);
+    const publicKeyBuffer = publicKey.export(publicKeyEncoding);
+    const privateKeyBuffer = privateKey.export(privateKeyEncoding);
+    const publicKeyBytes = getKeyMaterial(publicKeyBuffer);
+    const privateKeyBytes = getKeyMaterial(privateKeyBuffer);
+    return {
+      publicKey: publicKeyBytes,
+      secretKey: Buffer.concat([privateKeyBytes, publicKeyBytes])
+    };
+  },
+  async generateKeyPair({randomBytes}) {
+    return api.generateKeyPairFromSeed(randomBytes);
```

Comment:
Somewhere at the top of the file add:

```js
const generateRandomBytes = promisify(crypto.randomBytes);
```

```suggestion
  async generateKeyPair() {
    const seed = await generateRandomBytes(32);
    return api.generateKeyPairFromSeed(seed);
```

Reply:
node's `crypto.randomBytes` returns a Buffer, so no need to the promisify it looks like:

https://github.com/digitalbazaar/ed25519-verification-key-2020/pull/3/commits/c0ad8029f48bcff17eb8e19701b01a4b506a895d

Reply:
@aljones15 -- We're miscommunicating here. It returns a Buffer because it's synchronous. That means if the PRNG stalls (due to lack of entropy), you'll block the whole event loop (whole app) waiting on it. Now, I don't think there are any/many node.js PRNG implementations that don't use something like Fortuna that are able to avoid that in probably all cases. But it's something I've been burned by in the past and that's what I mention it. We're fine as is and we will promisify it later if it becomes a real problem.

Reply:
I see you promisified in another commit. Resolving this. Thanks!